pennylane_ionq/device.py

```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
 
-        return self._prob_array
+        Returns:
+            array[float]: list of the probabilities
+        """
+        try:
+            self._samples = self.generate_samples()
+            result = super().estimate_probability(wires, shot_range, bin_size)
+            self._samples = None
+            return result
+        except TypeError as e:
+            self._samples = None
+            e.args = ("When invoking QubitDevice estimate_probability() method after submitting circuits using any other method than QubitDevice.execute() method, current circuit index should be set via set_current_circuit_index() method pointing to index of the circuit which is targeted out of one or potentially several circuits submitted in a batch.", *e.args)
+            raise
+
+    def sample(
+        self,
+        observable,
+        shot_range=None,
+        bin_size=None,
+        counts=False,
+    ):
+        """Return samples of an observable. Overwrites the method in QubitDevice class
+        to accomodate batch submission of circuits. When invoking this method after submitting 
+        circuits using any other method than QubitDevice.execute() method, current circuit 
+        index should be set via set_current_circuit_index() method pointing to index of the 
+        circuit which is targeted out of one or potentially several circuits submitted in a batch.
+        This is not very good design but the alternative would be to create huge amounts of 
+        code duplication with respect to the Pennylane base code.
+
+        Args:
+            observable (Observable): the observable to sample
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+            counts (bool): whether counts (``True``) or raw samples (``False``)
+                should be returned
+            circuit_index (int): index of circuit in case of batch circuit submission
+
+        Raises:
+            EigvalsUndefinedError: if no information is available about the
+                eigenvalues of the observable
+
+        Returns:
+            Union[array[float], dict, list[dict]]: samples in an array of
+            dimension ``(shots,)`` or counts
+        """
+        try:
+            self._samples = self.generate_samples()

```

Comment:
Why generate `self._samples` again, and then set them to `None` after calculating the result? Didn't both `execute` and `batch_execute` already do `self._samples = self.generate_samples()`? Of course if we clear them here, we can't use them for the next measurement, but why clear them here instead of clearing them in `reset`?